stacks/api/bucket/bucket.controller.ts

```diff
@@ -262,4 +266,73 @@ export class BucketController {
       });
     });
   }
+
+  async clearRemovedFields(
+    bucketDataService: BucketDataService,
+    previousSchema: Bucket,
+    currentSchema: Bucket
+  ) {
+    let removedKeys = this.findRemovedKeys(
+      previousSchema.properties,
+      currentSchema.properties,
+      [],
+      ""
+    );
+    if (removedKeys.length < 1) return;
+
+    let unsetFields = removedKeys.reduce((pre, acc: any, index, array) => {
+      acc = {...pre, [array[index]]: ""};
+      return acc;
+    }, {});
+
+    await bucketDataService.updateMany(previousSchema._id, {}, {$unset: unsetFields});
+  }
+
+  findRemovedKeys(previousSchema: any, currentSchema: any, removedKeys: string[], path: string) {
+    for (const key of Object.keys(previousSchema)) {
+      if (!currentSchema.hasOwnProperty(key)) {
+        removedKeys.push(path ? `${path}.${key}` : key);
+        //we dont need to check child keys of this key anymore
+        continue;
+      }
+      if (this.isObject(previousSchema[key]) && this.isObject(currentSchema[key])) {
+        this.findRemovedKeys(
+          previousSchema[key].properties,
+          currentSchema[key].properties,
+          removedKeys,
+          path ? `${path}.${key}` : key
+        );
+      } else if (this.isArray(previousSchema[key]) && this.isArray(currentSchema[key])) {
+        this.addArrayPattern(
+          previousSchema[key].items,
+          currentSchema[key].items,
+          removedKeys,
+          path ? `${path}.${key}` : key
+        );
+      }
+    }
+    return removedKeys;
+  }
+
+  addArrayPattern(previousSchema: any, currentSchema: any, removedKeys: string[], path: string) {
+    path = `${path}.$[]`;
+    if (this.isArray(previousSchema) && this.isArray(currentSchema)) {
+      this.addArrayPattern(previousSchema.items, currentSchema.items, removedKeys, path);
+    } else if (this.isObject(previousSchema) && this.isObject(currentSchema)) {
+      this.findRemovedKeys(previousSchema.properties, currentSchema.properties, removedKeys, path);
+    }
+  }
+
+  isObject(schema: any) {
+    return schema &&
+      schema.type == "object" &&
+      schema.properties &&
+      Object.keys(schema.properties).length > 0
```

Comment:
Why do we check for the presence of properties?

Reply:
Actually it was for `Object.keys(schema.properties).length` might throw an error like `Cannot convert undefined or null to object` when the properties undefined. Is it possible to be undefined?

Reply:
I'm not sure whether it can be or not.  I don't get that why we even check for properties? 
If its type is 'object' then it should be treated as an object right?

Reply:
Of course, they should be like you said. But it's a recursive method and I didn't want to take the risk of interrupting that logic with this kind of errors. Our client-side can't send this kind of mistaken schema if I'm not wrong?

Reply:
Yes, it would send a schema whose properties property has '{}' as a value when there is no property. 

Also, note that ignoring a possible error may lead to unpredictable and unforeseen bugs.

Reply:
Okay, it looks better I remove this then, thanks for the suggestion.